app/page.tsx

```diff
@@ -0,0 +1,128 @@
+// @@@SNIPSTART typescript-next-oneclick-page-start
+'use client'
+import Head from 'next/head';
+import React from 'react';
+import { v4 as uuid4 } from 'uuid';
+import 'react-toastify/dist/ReactToastify.css';
+// @@@SNIPEND
+
+// @@@SNIPSTART typescript-next-oneclick-page-vars
+
+interface ProductProps {
+  product: {
+    id: number;
+    name: string;
+    price: string;
+  };
+}
+
+const products = [
+  {
+    id: 1,
+    name: 'PDF Book',
+    price: '$49',
+  },
+  {
+    id: 2,
+    name: 'Kindle Book',
+    price: '$49',
+  },
+];
+
+type ITEMSTATE = 'NEW' |  'ORDERED' | 'ERROR';
+// @@@SNIPEND
+
+// @@@SNIPSTART typescript-next-oneclick-page-product
+const Product: React.FC<ProductProps> = ({ product }) => {
+  const itemId = product.id;
+  const [state, setState] = React.useState<ITEMSTATE>('NEW');
+  const stateRef = React.useRef<ITEMSTATE>();
```

Comment:
Slight nit but can just import useState, useRef if you wanted to, but nbd at all. IE:

```
import React, { useState, useRef  } from 'react';

const [state, setState] = useState<ITEMSTATE>('NEW');
```